# Exploring eBay Car Sales Data

In this guided project, we'll work with a dataset of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website.

The aim of this project is to clean the data and analyse the included used car listings.

## The Dataset

We've made a few modifications from the original dataset (which you can find it [here](https://data.world/data-society/used-cars-data)):
- We sampled 50,000 data points from the full dataset, to ensure the code runs quickly
- We dirtied the data set a bit to more closely resemble what you would expect from a scraped dataset

The data dictionary provided with data is as follows:

- `dateCrawled`: When this ad was first crawled. All field-values are taken from this date
- `name`: Name of the car
- `seller`: Whether the seller is private or a dealer
- `offerType`: The type of listing
- `price`: The pricee on the ad to sell the car
- `abtest`: Whether the listing is included in an A/B test
- `vehicleType`: The vehicle type
- `yearOfRegistration`: The year in which the car was first registered
- `gearbox`: The transmission type
- `powerPS`: The power of the car in PS
- `model`: The car model name
- `kilmoeter`: How many kilometers the car has driven
- `monthOfRegistration`: The month in which the car was first registered
- `fuelType`: What type of fuel the car uses
- `brand`: The brand of the car
- `notRepairedDamage`: If the car has a damage which is not yet repaired
- `dateCreated`: The date on which the eBay listing was created
- `nrOfPictures`: The number of pictures in the ad
- `postalCode`: The postal code for the location of the vehicle
- `lastSeenOnline`: When the crawler saw this ad last online



## Reading in the dataset

In [3]:
import pandas as pd
import numpy as np
autos = pd.read_csv('C:/Users/Irene Lin/Desktop/autos.csv', encoding='Latin-1')

In [4]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [5]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

We can see that 5 columns have missing values, but none of them has more than 20% missing values. 5 columns own integer values and the rest of columns own string values.

We notice that the column names use camelcase instead of Python's preferred snakecase. So let's convert the column names to snakecase and reword some of the column names based on the data dictionary to be more descriptive.

In [6]:
print(autos.columns)

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')


In [7]:
camel_to_snake={
    'dateCrawled':'ad_crawled', 
    'name': 'name',
    'seller': 'seller',
    'price': 'price',
    'abtest': 'abtest',
    'gearbox': 'gearbox',
    'model': 'model',
    'odometer': 'odo_meter',
    'brand': 'brand',
    'offerType': 'offer_type', 
    'vehicleType' : 'vehicle_type', 
    'yearOfRegistration' : 'registration_year',
    'powerPS' : 'ps_power',
    'monthOfRegistration' : 'registration_month', 
    'fuelType' : 'fuel_type', 
    'notRepairedDamage' : 'unrepaired_damage', 
    'dateCreated' : 'ad_created', 
    'nrOfPictures' : 'picture_number', 
    'postalCode' : 'postal_code',
    'lastSeen' : 'last_seen'
}
autos.columns=autos.columns.map(camel_to_snake)

In [8]:
print(autos.columns)

Index(['ad_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'ps_power', 'model',
       'odo_meter', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'picture_number', 'postal_code',
       'last_seen'],
      dtype='object')


In [9]:
autos.head()

ad_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  ps_power   model  odo_meter  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  picture_number  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

In [10]:
autos.describe(include='all')

ad_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-30 19:48:02  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      ps_power  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

        odo_meter  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  picture_number   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

We notice that the columns __seller__ and __offer_type__ have mostly one value (49999 out of 50000), so they are candidates to be dropped.

Some important numeric data (such as __price__ and __odometer__) are stored as text, which needs to be cleaned in the next stage.

In [11]:
autos['price'] = autos['price'].str.replace('$','')
autos['price'] = autos['price'].str.replace(',','')
autos['price'] = autos['price'].astype(int)
autos.rename({'price':'price_dollar'}, axis=1, inplace=True)

autos['odo_meter'] = autos['odo_meter'].str.replace('km','')
autos['odo_meter'] = autos['odo_meter'].str.replace(',','')
autos['odo_meter'] = autos['odo_meter'].astype(int)
autos.rename({'odo_meter':'odometer_km'}, axis=1, inplace=True)

<ipython-input-11-55fdb9fd007b>:1: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will*not* be treated as literal strings when regex=True.
  autos['price'] = autos['price'].str.replace('$','')


Let's explore __price__ and __odometer__ more.

In [12]:
autos['price_dollar'].unique().shape

(2357,)

In [13]:
autos['price_dollar'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price_dollar, dtype: float64

It looks strange that minimal price is zero.

In [14]:
autos['price_dollar'].value_counts().sort_index(ascending=False).head(20)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: price_dollar, dtype: int64

In [15]:
autos['price_dollar'].value_counts().sort_index(ascending=True).head(20)

0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
20       4
25       5
29       1
30       7
35       1
Name: price_dollar, dtype: int64

Entries with over 350,000 dollars are seen as outliers because there is a big jump.

Considering eBay is an auction site, there could be cars with an opening bid of $1. So we remove the rows not with price between 1 and 350,000 dollars.

In [16]:
autos = autos[autos['price_dollar'].between(1,350000)]
autos['price_dollar'].describe()

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price_dollar, dtype: float64

## Exploring the date columns

In [17]:
autos['odometer_km'].unique().shape

(13,)

In [18]:
autos['odometer_km'].describe()

count     48565.000000
mean     125770.101925
std       39788.636804
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [19]:
autos['odometer_km'].value_counts().sort_index().head(20)

5000        836
10000       253
20000       762
30000       780
40000       815
50000      1012
60000      1155
70000      1217
80000      1415
90000      1734
100000     2115
125000     5057
150000    31414
Name: odometer_km, dtype: int64

It doesn't make any sense to see any value as outliers.

In [20]:
autos['ad_crawled'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: ad_crawled, dtype: float64

Crawling date is from March 5 to April 7, roughly over a month. And the distribution of listings crawled on each day is fairly even.

In [21]:
autos['ad_created'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
                ...   
2016-04-03    0.038855
2016-04-04    0.036858
2016-04-05    0.011819
2016-04-06    0.003253
2016-04-07    0.001256
Name: ad_created, Length: 76, dtype: float64

There seems to be a large variety of ad created dates. The oldest one is from June 2015.

In [22]:
autos['last_seen'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: last_seen, dtype: float64

Almost half listing was seen in the last three days (April 5 to 7).

In [23]:
autos['registration_year'].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

Because a car can't be first registered after the listing was seen, any vehicle with a registration year above 2016 is definitely inaccurate. Determining the earliest valid year is more difficult. Realistically, it could be somewhere in the first few decades of the 1900s.

Let's count the number of listings with cars that fall outside the 1900 - 2016 interval and see if it's safe to remove those rows entirely, or if we need more custom logic.

In [24]:
1- autos['registration_year'].between(1900,2016).sum() / len(autos['registration_year'])

0.03879336971069702

Less than 4% values are outside the interval of 1900 and 2016, so it's safe to remove those rows entirely.

In [25]:
autos = autos[autos['registration_year'].between(1900,2016)]
autos['registration_year'].value_counts(normalize=True)

2000    0.067608
2005    0.062895
1999    0.062060
2004    0.057904
2003    0.057818
          ...   
1929    0.000021
1931    0.000021
1938    0.000021
1939    0.000021
1952    0.000021
Name: registration_year, Length: 78, dtype: float64

The most 3 common registration years of cars are 2000, 2005 and 1999. The least 5 common years are 1938, 1948, 1927, 1931 and 1952.

## Exploring price by brand

In [26]:
autos['brand'].value_counts(normalize=True)

volkswagen        0.211264
bmw               0.110045
opel              0.107581
mercedes_benz     0.096463
audi              0.086566
ford              0.069900
renault           0.047150
peugeot           0.029841
fiat              0.025642
seat              0.018273
skoda             0.016409
nissan            0.015274
mazda             0.015188
smart             0.014160
citroen           0.014010
toyota            0.012703
hyundai           0.010025
sonstige_autos    0.009811
volvo             0.009147
mini              0.008762
mitsubishi        0.008226
honda             0.007840
kia               0.007069
alfa_romeo        0.006641
porsche           0.006127
suzuki            0.005934
chevrolet         0.005698
chrysler          0.003513
dacia             0.002635
daihatsu          0.002506
jeep              0.002271
subaru            0.002142
land_rover        0.002099
saab              0.001649
jaguar            0.001564
daewoo            0.001500
trabant           0.001392
r

Let's select those that have over 5% of the total values, namely __volkswagen, bmw, opel, mercedes_benz, audi and ford__. 

In [27]:
selected_brand = autos['brand'].value_counts(normalize=True).index[:6]

meanprice_by_brand = {}

for b in selected_brand:
    mean_price = autos.loc[autos['brand'] == b, 'price_dollar'].mean()
    meanprice_by_brand[b] = int(mean_price)
    
meanprice_by_brand

{'volkswagen': 5402,
 'bmw': 8332,
 'opel': 2975,
 'mercedes_benz': 8628,
 'audi': 9336,
 'ford': 3749}

We can observe that in the top 6 brands, there's a distinct price gap:
- Audi, BMW, and Mercedes Benz are luxury brands
- Ford and Opel are less expensive
- Volkswagen is in between

In [28]:
# mean mileage of top 6 brands

selected_brand = autos['brand'].value_counts(normalize=True).index[:6]

mileage_by_brand = {}

for b in selected_brand:
    mean_mileage = autos.loc[autos['brand'] == b, 'odometer_km'].mean()
    mileage_by_brand[b] = int(mean_mileage)
    
mileage_by_brand

{'volkswagen': 128707,
 'bmw': 132572,
 'opel': 129310,
 'mercedes_benz': 130788,
 'audi': 129157,
 'ford': 124266}

In [29]:
# convert both dictionaries to series, then to one dataframe

price_series = pd.Series(meanprice_by_brand)
mileage_series = pd.Series(mileage_by_brand)

price_mileage_df = pd.DataFrame(price_series, columns=['mean_price'])
price_mileage_df['mean_mileage'] = mileage_series
price_mileage_df

mean_price  mean_mileage
volkswagen           5402        128707
bmw                  8332        132572
opel                 2975        129310
mercedes_benz        8628        130788
audi                 9336        129157
ford                 3749        124266

From the dataframe above representing mean price and mean mileage of top 6 brands, we can see that mean mileage doesn't vary a lot, in spite of varities in mean price.

## Translating german words to english

Some categorical data use german words, so we need to translate them and map the values to their english counterparts.

In [30]:
autos['vehicle_type'].value_counts()

limousine     12598
kleinwagen    10585
kombi          8930
bus            4031
cabrio         3016
coupe          2462
suv            1965
andere          390
Name: vehicle_type, dtype: int64

In [31]:
vehicle_translate = {
    'limousine' : 'limousine',
    'kleinwagen' : 'small car',
    'kombi' : 'combi',
    'bus' : 'bus',
    'cabrio' : 'convertible',
    'coupe' : 'coupe',
    'suv' : 'suv',
    'andere' :'other'
}
autos['vehicle_type'] = autos['vehicle_type'].map(vehicle_translate)

In [32]:
autos['gearbox'].value_counts()

manuell      34715
automatik     9856
Name: gearbox, dtype: int64

In [33]:
gearbox_translate = {
    'manuell' : 'manually',
    'automatik' : 'automatic'
}
autos['gearbox'] = autos['gearbox'].map(gearbox_translate)

In [34]:
autos['fuel_type'].value_counts()

benzin     28540
diesel     14032
lpg          649
cng           71
hybrid        37
elektro       19
andere        15
Name: fuel_type, dtype: int64

In [35]:
fueltype_translate = {
    'benzin' : 'petrol',
    'diesel' : 'diesel',
    'lpg' : 'lpg',
    'cng' : 'cng',
    'hybrid' : 'hybrid',
    'elektro' : 'electro',
    'andere' : 'other'
}
autos['fuel_type'] = autos['fuel_type'].map(fueltype_translate)

In [46]:
autos['unrepaired_damage'].value_counts()

nein    33834
ja       4540
Name: unrepaired_damage, dtype: int64

In [47]:
damage_translate = {
    'nein' : 'no',
    'ja' : 'yes'
}
autos['unrepaired_damage'] = autos['unrepaired_damage'].map(damage_translate)

## Patterns of average prices based on the mileage

In [43]:
mileage_group = autos['odometer_km'].value_counts().index
mileage_mean_price = {}

for m in mileage_group:
    mean_price = autos.loc[ autos['odometer_km']==m, 'price_dollar'].mean()
    mileage_mean_price[m] = int(mean_price)
    
mileage_mean_price

{150000: 3767,
 125000: 6214,
 100000: 8132,
 90000: 8465,
 80000: 9721,
 70000: 10927,
 60000: 12385,
 50000: 13812,
 40000: 15499,
 5000: 8873,
 30000: 16608,
 20000: 18448,
 10000: 20550}

We can see the pattern that mean price decreases as used mileage increases, which follows common sense.

## Falling price due to car damage

We notice that there is a column which indicates whether a car has unrepaired damage. Let's explore how the damage affects the price.

In [50]:
autos.loc[ autos['unrepaired_damage'] == 'no', 'price_dollar'].mean()


7164.033102796004

In [51]:
autos.loc[ autos['unrepaired_damage'] == 'yes', 'price_dollar'].mean()

2241.146035242291

It seems that damage has an important effect on price.

## Finding the most common brand/model combinations

In [53]:
autos[['brand', 'model']].value_counts()

brand       model     
volkswagen  golf          3707
bmw         3er           2615
volkswagen  polo          1609
opel        corsa         1592
volkswagen  passat        1349
                          ... 
ford        b_max            1
rover       rangerover       1
bmw         i3               1
rover       discovery        1
audi        200              1
Length: 290, dtype: int64

So the most common brand/model combinations are __volkswagen/golf__ and __bmw/3er__.